src/type_manager.cpp

```diff
@@ -70,89 +66,127 @@ void TypeManager::register_engine_singletons(jni::Env& p_env, jni::JObjectArray&
 #endif
             delete ref->kotlin_class;
+            ref->kotlin_class = nullptr;
         }
 
         // We only add them back if they are in use, otherwise we let the Script die.
         if (!ref->placeholders.is_empty()) { scripts.push_back(ref); }
     }
 
 #else
+    // This part handles the first and only loading of script happening in exports.
+    // It assumes that all scripts are properly built into the .jar and won't be reloaded.
 #ifdef DEBUG_ENABLED
-    JVM_ERR_FAIL_COND_MSG(user_scripts.size() != 0, "Kotlin scripts are being initialized more than once.");
+    JVM_ERR_FAIL_COND_MSG(named_user_scripts.size() != 0, "JVM scripts are being initialized more than once.");
 #endif
 
     for (KtClass* kotlin_class : classes) {
         Ref<KotlinScript> ref;
         ref.instantiate();
         ref->kotlin_class = kotlin_class;
+        ref->mode = KotlinScript::AccessMode::NAME;
+        ref->set_path(kotlin_class->compilation_time_relative_registration_file_path, true);
         scripts.push_back(ref);
 #ifdef DEV_ENABLED
         LOG_VERBOSE(vformat("Kotlin Script created: %s", kotlin_class->registered_class_name));
 #endif
     }
 #endif
 
-    for (Ref<KotlinScript> script : scripts) {
-        user_scripts.push_back(script);
-        user_scripts_map[script->get_global_name()] = script;
+    for (const Ref<KotlinScript>& script : scripts) {
+        named_user_scripts.push_back(script);
+        named_user_scripts_map[script->get_global_name()] = script;
+    }
+
+    // Now we deal with path script.
+    Vector<Ref<KotlinScript>> path_script_cache = path_user_scripts;
+    path_user_scripts.clear();
+    for (Ref<KotlinScript>& script : path_script_cache) {
+        String path = script->get_path();
+        // No need to delete the KotlinClass, it has already been done with the namedScript that shares it.
+        script->kotlin_class = nullptr;
+        if (filepath_to_name_map.has(path)) {
+            script->kotlin_class = named_user_scripts_map[filepath_to_name_map[path]]->kotlin_class;
+            path_user_scripts.push_back(script);
+        }
     }
 
 #ifdef TOOLS_ENABLED
-    for (Ref<KotlinScript> script : user_scripts) {
+    for (const Ref<KotlinScript>& script : named_user_scripts) {
+        // We have to delay the update_export. The engine is not fully initialized and scripts can cause undefined behaviors.
+        MessageQueue::get_singleton()->push_callable(callable_mp(script.ptr(), &KotlinScript::update_exports));
+    }
+    for (const Ref<KotlinScript>& script : path_user_scripts) {
         // We have to delay the update_export. The engine is not fully initialized and scripts can cause undefined behaviors.
         MessageQueue::get_singleton()->push_callable(callable_mp(script.ptr(), &KotlinScript::update_exports));
     }
 #endif
 }
 
-Ref<KotlinScript> TypeManager::create_placeholder_script(String p_path) {
+Ref<KotlinScript> TypeManager::create_script(const String& p_path, bool named) {

```

Comment:
This method is only called with hardcoded boolean in `JvmResourceFormatLoader`.
So `named` boolean can be a template argument to statically resolve the if at compile time and so avoid a branch at runtime.